content/assignments/minishell/checker/_test/run_test.sh

```diff
@@ -0,0 +1,244 @@
+export max_points=100
+TEST_TIMEOUT=30
+
+TEST_LIB=_test/test_lib.sh
+export CHECK_SKIP='parser.\(\(tab\|yy\).\(c\|h\)\|h\)$'
+
+# Load the lib functions.
+if ! [ -e "$TEST_LIB" ]; then
+	echo "Test library not found. Check \$TEST_LIB ($TEST_LIB)"
+	exit 1
+fi
+source "$TEST_LIB"
+
+export SHELL_PROMPT="> "
+
+# Change this of you want to keep the logs after execution.
+DO_CLEANUP=no
+
+BUFFERING_WRAPPER=""
+if command -v stdbuf &> /dev/null; then
+	# stdbuf may not be installed on cygwin.
+	BUFFERING_WRAPPER="stdbuf -i 0"
+fi
+
+USE_VALGRIND=yes
+VALGRIND_LOG="valgrind.log"
+
+# ----------------- Init and cleanup tests ----------------------------------- #
+
+# Initializes a test.
+init_test()
+{
+
+	if ! [ -e "$exec_name" ]; then
+		echo "$exec_name not found! Cannot run the test"
+		exit 1
+	fi
+
+	if [ "$USE_VALGRIND" = "yes" ]; then
+		which valgrind &> /dev/null
+		if [ $? -eq 0 ]; then
+			exec_name="valgrind --leak-check=full \
+			                    --show-leak-kinds=all \
+					    --child-silent-after-fork=yes \
+					    --track-fds=yes \
+					    --log-file=$VALGRIND_LOG \
+					    $exec_name"
+		else
+			echo "valgrind is not installed."
+		fi
+	fi
+
+	# Generates random file as test file.
+	TEST_NAME=$(printf "test_%02d" "$test_index")
+	OUT_DIR="${TEST_NAME}_out"
+	REF_DIR="${TEST_NAME}_ref"
+	IN_FILE="${TEST_NAME}.in"
+	REF_FILE="${TEST_NAME}.ref"
+	OUT_FILE="${TEST_NAME}.out"
+
+	local input_basename="${INPUT_DIR}/${TEST_NAME}"
+	[ -f "${input_basename}.${OS_PLATFORM}.txt" ] && \
+		input_basename="${input_basename}.${OS_PLATFORM}" || \
+		input_basename="${input_basename}"
+
+	# Copy input files in the testing directory.
+	cp "${input_basename}.txt" "${MAIN_TEST_DIR}/${IN_FILE}" &> "$LOG_FILE"
+	# Copy additional files, if they exist (see test 04).
+	# cp ${input_basename}?*.txt ${MAIN_TEST_DIR} &> $LOG_FILE
+	# Copy any reference files, if they exist (see test 18).
+	cp "${REFS_DIR}/${TEST_NAME}.ref" "${MAIN_TEST_DIR}" &> "$LOG_FILE"
+
+	# Export mini-shell in path.
+	PATH="$PATH:$(pwd)"
+	cd "${MAIN_TEST_DIR}" &> "$LOG_FILE" || exit 1
+
+	[ -d "${OUT_DIR}" ] && rm -rf "${OUT_DIR}"
+	[ -d "${REF_DIR}" ] && rm -rf "${REF_DIR}"
+}
+
+# Cleanups a test.
+cleanup_test()
+{
+	if [ "$DO_CLEANUP" = "yes" ]; then
+		[ -d "${OUT_DIR}" ] && rm -rf "${OUT_DIR}"
+		[ -d "${REF_DIR}" ] && rm -rf "${REF_DIR}"
+		[ -e "${IN_BASH}" ] && rm "${IN_BASH}"
+		[ -e "${IN_MINI}" ] && rm "${IN_MINI}"
+	fi
+}
+
+# Initializes the whole testing environment.
+# Should be the first test called.
+init_world()
+{
+	[ -d "$MAIN_TEST_DIR" ] && rm -rf "$MAIN_TEST_DIR"
+	mkdir -p "${MAIN_TEST_DIR}"
+	print_header "Testing - mini-shell"
+}
+
+# Cleanups the whole testing environment.
+# Should be the last test called.
+cleanup_world()
+{
+	[ "$DO_CLEANUP" = "yes" ] && rm -rf "$MAIN_TEST_DIR"
+}
+
+# ----------------- Test Suite ----------------------------------------------- #
+
+# Specific test example.
+execute_cmd()
+{
+	EXEC=$1
+	INPUT=$2
+	OUTPUT=$3
```

Comment:
The arguments should normally be escaped. For the first argument (`EXEC`) this may be an issue since you need having multiple arguments when running `valgrind`. I think it would be easier to have a `GEN_REF` flag (or similar) and use a function to run the correct command - e.g.,
```bash
eval_command() {
    GEN_REF="$1"
    shift 1 # drop the first argument
    if [ "${GEN_REF}" = "yes" ]; then
       "$ref_name"
    elif [ "${USE_VALGRIND}" = "yes" ]; then
       # valgrind and all required arguments
    else
       "${exec_name}"
    fi
}
```